src/reformatters/common/binary_rounding.py

```diff
@@ -0,0 +1,87 @@
+from typing import Final
+
+import numpy as np
+import numpy.typing as npt
+
+# Type aliases
+FloatArray = npt.NDArray[np.float32]
+IntArray = npt.NDArray[np.uint32]
+
+# IEEE 754 binary32 constants
+MANTISSA_BITS: Final[int] = 23
+MANTISSA_MASK: Final[np.uint32] = np.uint32((1 << MANTISSA_BITS) - 1)
+EXPONENT_MASK: Final[np.uint32] = np.uint32(((1 << 8) - 1) << (32 - 8 - 1))
+SIGN_MASK: Final[np.uint32] = np.uint32(1 << 31)
+
+
+def round_float32(value: FloatArray, keep_mantissa_bits: int) -> FloatArray:
+    """Round float32 values to keep a specific number of mantissa bits.
+
+    Vectorized implementation that works on numpy float32 arrays.
+
+    Args:
+        value: The float32 array to round
+        keep_mantissa_bits: Number of mantissa bits to keep (1-23)
+
+    Returns:
+        Rounded float32 array
+    """
+    if keep_mantissa_bits >= MANTISSA_BITS:
+        return value
+
+    # Early return for any NaN or Inf values
+    if np.any(~np.isfinite(value)):
+        return value.copy()
```

Comment:
This should store a mask of nan values to re-apply later, so we can still round if some values are nan